### ISSUE TITLE
Block access to system indices from REST APIs

### ISSUE THREAD

**Issue by andrross** [opensearch-project/OpenSearch#7936]
Access to system indices via the REST APIs [has been deprecated](https://github.com/opensearch-project/OpenSearch/commit/5c8b0662df09a5d7a53fd0d98b1d37b1d831ddbc) since OpenSearch 1.x (and ES 7.10 before that). This issue is to complete the deprecation and block access to system indices via the REST API. This is obviously a breaking change and can only be completed in the next major version (3.0).

**Comment by dbwiddis** [comment]
What will be the supported replacement method for accessing these indices?  

In particular, many plugin integration tests clean up indices at the end of a test class in order to not impact another test class.  See https://github.com/search?q=org%3Aopensearch-project+wipeAllODFEIndices&type=code

```java
    @After
    protected void wipeAllODFEIndices() throws IOException {
        Response response = client().performRequest(new Request("GET", "/_cat/indices?format=json&expand_wildcards=all"));
        // parse response
        for (Object index : parser.list()) {
            Map<String, Object> jsonObject = (Map<String, Object>) index;
            String indexName = jsonObject.get("index").toString();
            if (!".opendistro_security".equals(indexName)) {
                Request request = new Request("DELETE", String.format(Locale.getDefault(), "/%s", indexName));
                // process
            }
        }
    }
```

**Comment by cwperks** [comment]
@dbwiddis OpenSearch already has system index protection by default in the default distribution of OpenSearch w/ the security plugin installed. (Strong) System Index Protection is provided by the security plugin which only allows system index access in 2 cases:

1. As the system making calls (i.e. blocks wrapped by `try (ThreadContext.StoredContext ctx = threadContext.stashContext()) { ... }` or just running in a fresh ThreadContext in general)
2. Calls as the superuser (making REST calls with the admin certificate)

Many plugins already account for this and in the snippet you provide above the `client()` would need to be configured with the admin certificate.

W/o the security plugin installed is a different story. W/o encryption, there isn't a foolproof way to secure system indices. At best, OpenSearch could require a certain HTTP Header to be present in the request to declare the intent to make a request on a system index, but since the requests are unencrypted you can't ensure the integrity of any secrets.

Check out this comment from @nibix here: https://github.com/opensearch-project/security/pull/4896#discussion_r1889683926

> I'd agree to this analysis, @cwperks .

> Actually, core already contains ground works for a protection of system indices. It works like this:

> In case a REST request is issues against a system index, the following message will be logged: this request accesses system indices: [{}], but in a future major version, direct access to system indices will be prevented by default
Exception: The HTTP header X-opensearch-product-origin is present in the REST request
Other exception: The RestHandler implementation overrides allowSystemIndexAccessByDefault() and returns true.
See this commit for details: https://github.com/opensearch-project/OpenSearch/commit/5c8b0662df09a5d7a53fd0d98b1d37b1d831ddbc

> These concepts still come from the ES times. If I understand it correctly, the goal was to replace the warning message by an error as a breaking change with a major release.

> IMHO, this concept makes totally sense for core. As @cwperks has pointed out, core has no safe means of authentication. What core can do is a kind of cooperative check: If a client assures that they need to perform operations on a system index, they are allowed to. If the client signals no concrete intent to operate on system index, they are not allowed to. This avoids unwanted operations in case of bugs on side of the client - be it an overly broad index pattern or a similar case. Of course, it does not provide real security. But that just takes into account the fact that core has no means to perform a secure authentication.

> IMHO, a good strategy would be to build on the present ground works and to complete them. That would be also useful from a code hygiene point of view, as otherwise that existing code would be there without real purpose.

> Relevant code pieces:

> https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L359-L378

> https://github.com/opensearch-project/OpenSearch/blame/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/rest/RestController.java#L374-L381

> https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L767-L774

**Comment by dbwiddis** [comment]
> Many plugins already account for this and in the snippet you provide above the client() would need to be configured with the admin certificate.

The `client()` in this case is inherited from [`OpenSearchTestCase`](https://github.com/opensearch-project/OpenSearch/blob/main/test/framework/src/main/java/org/opensearch/test/rest/OpenSearchRestTestCase.java).  Looking around in there, it already has a `wipeAllIndices()` call but these workaround patterns are typically designed where we want to exclude one.

But I notice that class uses `adminClient()` for this indices-wiping step and ignores the warnings.  This is [built with the same settings](https://github.com/opensearch-project/OpenSearch/blob/77e91c2fe11b86ec682d1d937bec0f2d000392f1/test/framework/src/main/java/org/opensearch/test/rest/OpenSearchRestTestCase.java#L808) by default so seems to not just be an easy fix.  But I assume the correct thing is for plugins to override that client and add the `X-opensearch-product-origin` header?

To be clear, while the above explanation is very useful, my primary goal here is to update my plugin to not break with 3.0, and all my integ tests (run without security plugin) currently produce this deprecation warning on the delete step.

**Comment by cwperks** [comment]
> To be clear, while the above explanation is very useful, my primary goal here is to update my plugin to not break with 3.0, and all my integ tests (run without security plugin) currently produce this deprecation warning on the delete step.

The deprecation message is coming from a [commit](https://github.com/opensearch-project/OpenSearch/commit/5c8b0662df09a5d7a53fd0d98b1d37b1d831ddbc) before the fork. I agree with @nibix that there should be an effort to clean this up bc the message is there in perpetuity otherwise.

> IMHO, a good strategy would be to build on the present ground works and to complete them. That would be also useful from a code hygiene point of view, as otherwise that existing code would be there without real purpose.

When using the security plugin, plugins should configure the admin client with the admin certificate ([example](https://github.com/opensearch-project/job-scheduler/blob/a6595929fefc650e4c34219d209c412e43be8259/sample-extension-plugin/src/test/java/org/opensearch/jobscheduler/sampleextension/SampleExtensionIntegTestCase.java#L82-L92)). Using the demo security config, `kirk.pem` and `kirk-key.pem` contain the admin cert and key. Specifically the [`plugins.security.authcz.admin_dn`](https://github.com/opensearch-project/security/blob/472f7c7979502606ccb0707908ce5f689dde6164/config/opensearch.yml.example#L30-L33) setting contains a list of distinguished names corresponding to admin certificates. You can find an example cURL request on the documentation website about [System Indexes](https://opensearch.org/docs/latest/security/configuration/system-indices/). 

Plugins sometimes checking the demo certs to their repositories, or they can obtain them from the security repo like this: https://github.com/opensearch-project/job-scheduler/blob/main/build.gradle#L54-L65

**Comment by whitepiratebaku** [comment]
Hi, usually we was using scripts to control the number of replicas for system indices which used API calls. Is there better way to control replicas count for internal indices? it is like in single node case (at least with 2.19) we was getting replica shards created for indices but since it was 1 node setup we was getting yellow state always.

I see [here](https://forum.opensearch.org/t/number-of-replicas-in-opensearch-single-node-setup/12523/3?utm_source=chatgpt.com) it is complained that _**cluster.default_number_of_replicas**_ also not covering internal indices, it is still the case?

**Comment by kumargu** [comment]
> IMHO, this concept makes totally sense for core. As @cwperks has pointed out, core has no safe means of authentication. What core can do is a kind of cooperative check: If a client assures that they need to perform operations on a system index, they are allowed to. If the client signals no concrete intent to operate on system index, they are not allowed to. This avoids unwanted operations in case of bugs on side of the client - be it an overly broad index pattern or a similar case. Of course, it does not provide real security. But that just takes into account the fact that core has no means to perform a secure authentication.


But do we really need any authentication mechanisms? The core should just filter out system indices from any GET / PUT requests.

**Comment by cwperks** [comment]
@kumargu I agree that there should be a mechanism to enforce that system indices are hidden. Currently, it's up to plugins that declare system indices to create them as hidden but it should be enforced.

Still if they are enforced as hidden, it's still a cooperative check w/o security because a caller must pass ?expand_wildcards=all.

**Comment by kumargu** [comment]
> Still if they are enforced as hidden, it's still a cooperative check w/o security because a caller must pass ?expand_wildcards=all.


Yes. the hidden check is kinda useless. It just best efforts but very easy to break.

**Comment by nibix** [comment]
@kumargu 

> It just best efforts but very easy break.

Could you elaborate on why you consider it easy to break? If there are not any authentication mechanisms, there needs to be some way to also gain access to these indices, for admin or debug purposes.

**Comment by kumargu** [comment]
I think even though they are marked as hidden, they could be searched. 

```curl -XGET "http://localhost:9200/_cat/indices?expand_wildcards=hidden,open&v"```

```
curl -X GET "http://localhost:9200/_search?expand_wildcards=hidden,open" -H 'Content-Type: application/json' -d '{
  "query": {
    "match_all": {}
  }
}'```


Agreed that we would need admin to gain access for debugging, but do we need security plugin for that? The cert based verification for admins to query these special indices could be baked out of security plugin?  A simpler mechanism could be to add a debug flag to be able to see these indices.  The debug flag cannot be passed as request params, someone (admin) should have access to underlying host to be able to toggle the flag.

**Comment by nibix** [comment]
@kumargu 

> A simpler mechanism could be to add a debug flag to be able to see these indices.  The debug flag cannot be passed as request params, someone (admin) should have access to underlying host to be able to toggle the flag. 

There is already code in core that has the goal of implementing a functionality very similar to what you are describing. A HTTP header is being used to indicate that access to system indices shall be granted (not only for debug purposes, but also for accessing the `.kibana` index by Dashboards).

Work on this functionality was already begun with in the ES times; that's why the knowledge about the feature kind of disappeared. More details here:

https://github.com/opensearch-project/security/pull/4896#discussion_r1889683926

**Comment by kumargu** [comment]
Got it @nibix. I'll read through the conversation. 

Should we even avoid passing the flag via HTTP headers? Our goal is to prevent REST access, so any form of usage via header, payload, uri -- should be avoided.

**Comment by nibix** [comment]
But at the moment, REST is the only core interface we have. If an admin wants to view the content of a system index for debugging purposes, they can only use the REST API. Likewise, Dashboards also uses the REST API to access its internal index.

**Comment by cwperks** [comment]
One such improvement on the current model could be to extend system index protection to also be enforced in `sslOnly` mode of the security plugin. W/o EIT, the best we can do is a cooperative check for protection.
